DEV: add index lifecycle page (#1064)
redis/docs: content/develop/interact/search-and-query/best-practices/index-mgmt-best-practices.md

- [`GEO`]({{< relref "/develop/interact/search-and-query/basic-constructs/field-and-type-options#geo-fields" >}}): use `GEO` for geospatial coordinates (e.g., latitude/longitude).
- [`GEOSHAPE`]({{< relref "/develop/interact/search-and-query/basic-constructs/field-and-type-options#geoshape-fields" >}}): use `GEOSHAPE` to represent locations as points, but also to define shapes and query the interactions between points and shapes (e.g., to find all points that are contained within an enclosing shape).
- [`VECTOR`]({{< relref "/develop/interact/search-and-query/basic-constructs/field-and-type-options#vector-fields" >}}): use `VECTOR` for high-dimensional similarity searches.

See [these pages]({{< relref "/develop/interact/search-and-query/query" >}}) for discussions and examples on how best to use these index types.

Next, simulate queries on a sample dataset to identify potential bottlenecks.
Use tools like [`FT.PROFILE`]({{< baseurl >}}/commands/ft.profile) to analyze query execution and refine your schema if needed.

> For example, assign weights to `TEXT` fields for prioritizing results or use the `PREFIX` option of [`FT.CREATE`]({{< baseurl >}}/commands/ft.create) to limit indexing to specific key patterns. Note: you can use multiple `PREFIX` clauses when you create an index.

Reviewer: Minor: maybe put the note sentence in parentheses or say "Note that you can..."?

After creating the index, validate its performance with real queries and monitor usage with the available tools:

- [`FT.EXPLAIN`]({{< baseurl >}}/commands/ft.explain) and [`FT.EXPLAINCLI`]({{< baseurl >}}/commands/ft.explaincli) allow you to see how Redis Query Engine parses a given search query. `FT.EXPLAIN` returns a structured breakdown of the query execution plan, while `FT.EXPLAINCLI` presents a more readable, tree-like format for easier interpretation. These commands are useful for diagnosing query structure and ensuring it aligns with the intended logic.

> - [`FT.INFO `]({{< baseurl >}}/commands/ft.info) provides detailed statistics about an index, including the number of indexed documents, memory usage, and configuration settings. It helps in monitoring index growth, assessing memory consumption, and verifying index structure to detect potential inefficiencies.

Reviewer: The extra space is underlined in the link text.
```suggestion
- [`FT.INFO`]({{< baseurl >}}/commands/ft.info) provides detailed statistics about an index, including the number of indexed documents, memory usage, and configuration settings. It helps in monitoring index growth, assessing memory consumption, and verifying index structure to detect potential inefficiencies.
```

> - [`FT.PROFILE`]({{< baseurl >}}/commands/ft.profile) runs a query while capturing execution details, revealing query performance bottlenecks. It provides insights into processing time, key accesses, and filter application, making it a crucial tool for fine-tuning complex queries and optimizing search efficiency.

Reviewer: Again minor, but "capturing... revealing" reads a bit oddly to me. It's as if the "while" covers both the capturing and revealing and then you're expecting a third action at the end.
```suggestion
- [`FT.PROFILE`]({{< baseurl >}}/commands/ft.profile) runs a query while capturing execution details, which helps to reveal query performance bottlenecks. It provides insights into processing time, key accesses, and filter application, making it a crucial tool for fine-tuning complex queries and optimizing search efficiency.
```

Avoid over-indexing. Indexing every field increases memory usage and can slow down updates.
Only index the fields that are essential for your planned queries.

## Index creation
   - Use the [`FT.CREATE`]({{< baseurl >}}/commands/ft.create) command to define an index schema.
   - Assign weights to `TEXT` fields to prioritize certain fields in full-text search results.

>    - Leverage the `PREFIX` option to restrict indexing to keys with specific patterns.

Reviewer: Or maybe "Use the `PREFIX` option carefully/wisely..."?
```suggestion
   - Use the `PREFIX` option to restrict indexing to keys with specific patterns.
```

          ```bash
          FT.CREATE my_index ON HASH PREFIX 2 "user:" "order:" SCHEMA name TEXT age NUMERIC status TAG
          ```

          This approach enables searching across multiple entity types without needing separate indexes.

      - Instead of querying multiple indexes separately, you can search across related data structures using a single query. This is particularly helpful when data structures share common fields, such as searching both customer and vendor records under a unified contacts index.

      - Maintaining multiple indexes for similar data types can be inefficient in terms of memory and query performance. By consolidating data under one index with multiple prefixes, you reduce overhead while still allowing for distinct key organization.

      - If your data model evolves and new key patterns are introduced, using multiple `PREFIX` clauses from the start ensures future compatibility without requiring a full reindexing.
   - Data loading strategy: load data into Redis before creating an index when working with large datasets. Use the `ON HASH` or `ON JSON` options to match the data structure.

## Index aliasing

> Index aliasing allows you to assign an alias to an index. Aliases act as abstracted names for the underlying indexes, enabling applications to reference the alias instead of the actual index name. This approach simplifies schema updates and index management.

Reviewer: ```suggestion
Index aliases act as abstracted names for the underlying indexes, enabling applications to reference the alias instead of the actual index name. This approach simplifies schema updates and index management.
```

There are several use cases for index aliasing, including:

- Schema updates: when updating an index schema, create a new index and associate the same alias with it. This allows a seamless transition without requiring application-level changes.
- Version control: use aliases to manage different versions of an index. For example, assign the alias products to `products_v1` initially and later to `products_v2` when the schema evolves.
- Testing and rollback: assign an alias to a test index during staged deployments. If issues arise, quickly switch the alias back to the stable index.

Best practices for aliasing:

> - Default aliases: always create an alias for your indexes during initial setup, even if you don’t anticipate immediate schema changes.
> - Consistent naming conventions: use clear and descriptive alias names to avoid confusion (e.g., `users_current` or `orders_live`).

Reviewer: The other two bullet points begin with a verb.
```suggestion
- Use default aliases: always create an alias for your indexes during initial setup, even if you don’t anticipate immediate schema changes.
- Use consistent naming conventions: use clear and descriptive alias names to avoid confusion (e.g., `users_current` or `orders_live`).
```

    ```
 - Implement scripts to periodically verify document counts and query results. For example, in Python:

    ```python
    import re    
    def check_index_readiness(index_name, expected_docs):
        r = redis.StrictRedis(host='localhost', port=6379, decode_responses=True)
        info = r.execute_command('FT.INFO', index_name)
        num_docs = int(info[info.index('num_docs') + 1])
        return num_docs >= expected_d    
    if check_index_readiness('my_new_index', 100000):
        print("Index is fully populated!")
    else:
        print("Index is still populating...")
    ```

## Monitoring index performance

- Use the `FT.PROFILE` command to analyze query performance and identify bottlenecks.
- Regularly monitor memory usage with the [`INFO`]({{< baseurl >}}/commands/info) `memory` and `FT.INFO` commands to detect growth patterns and optimize resource allocation.

## Index maintenance

- If schema changes are required, create a new index with the updated schema and reassign the alias once the index is ready.
- Use [Redis key expiration]({{< relref "/develop/use/keyspace#key-expiration" >}}) to automatically remove outdated records and keep indexes lean.

> When is it appropriate to use [`FT.ALTER`]({{< baseurl >}}/commands/ft.alter) and when is it best to use aliasing?

Reviewer: Or maybe rephrase it so it isn't a question?
```suggestion
### When is it appropriate to use [`FT.ALTER`]({{< baseurl >}}/commands/ft.alter) and when is it best to use aliasing?
```

---
Title: Index management best practices for Redis Query Engine
alwaysopen: false
categories:
- docs
- develop
- stack
- oss
- kubernetes
- clients
linkTitle: RQE index management
weight: 3
---

> 

Author: Fixed in next push.

#### 1. Plan your indexes strategically

>    - Understand your query patterns: before creating indexes, analyze your expected query patterns to ensure indexes are optimized for performance.

Author: Fixed in next push.

   - Avoid over-indexing: indexing every field increases memory usage and can slow down updates. Only index fields essential for your queries.

>    - Choose appropriate index types: use the correct field types (`TEXT`, `TAG`, `NUMERIC`, `GEO`, or `VECTOR`) for your data to maximize efficiency.

Author: Fixed in next push.

#### 2. Index creation

>    - Atomic creation: use the `FT.CREATE` command to atomically define an index schema.

Author: Fixed in next push.

   - Field weighting: assign weights to `TEXT` fields to prioritize certain fields in full-text search results.

>    - Prefix optimization: leverage the `PREFIX` option to restrict indexing to keys with specific patterns.

Reviewer 2: Customers tend to get confused by prefixes and indices, even though it is a simple construct. Also when should they use multiple prefixes in an index verses a single one?

I think we could offer a bit more here for education!

Author: Fixed in next push.

   - Data loading strategy: load data into Redis before creating an index when working with large datasets. Use the `ON HASH` or `ON JSON` options to match the data structure.

#### 3. Index aliasing

>    - What is index aliasing?

Author: Fixed in next push.

     - Aliases act as abstracted names for indexes, allowing applications to reference the alias instead of the actual index name. This simplifies schema updates and index management.
   - Use cases for index aliasing:
     - Seamless schema updates: point the alias to a new index without changing application code.
     - Version control: assign aliases like `products_live` to track the active index version.
     - Testing and rollback: temporarily assign aliases to test indexes, and revert if needed.
   - How to manage aliases:
     - Assign an alias: `FT.ALIASADD my_alias my_index`
     - Update an alias: `FT.ALIASUPDATE my_alias new_index`
     - Remove an alias: `FT.ALIASDEL my_alias`

#### 4. Monitoring index population
   - Check document count:

>      - Use the `FT.INFO` command to monitor the `num_docs` field, ensuring all expected documents are indexed.

Reviewer 2: There is also the `indexing` attribute that is often useful from ft.info

Author: Fixed in next push.

>      - Use the `FT.INFO` command to monitor the `num_docs` field, ensuring all expected documents are indexed.

Author: Fixed in next push.

     - Example:
       ```bash
       FT.INFO my_new_index
       ```
   - Run test queries:
     - Validate data with sample queries to ensure proper indexing:
       ```bash
       FT.SEARCH my_new_index "*"
       ```

>    - Query profiling:

Author: Fixed in next push.

     - Use `FT.PROFILE` to analyze query plans and validate performance:
       ```bash
       FT.PROFILE my_new_index SEARCH QUERY "your_query"
       ```
   - Automate checks:
     - Implement scripts to periodically verify document counts and query results. For example, in Python:
       ```python
       import redis

       def check_index_readiness(index_name, expected_docs):
           r = redis.StrictRedis(host='localhost', port=6379, decode_responses=True)
           info = r.execute_command('FT.INFO', index_name)
           num_docs = int(info[info.index('num_docs') + 1])
           return num_docs >= expected_docs

       if check_index_readiness('my_new_index', 100000):
           print("Index is fully populated!")
       else:
           print("Index is still populating...")
       ```

> #### 5. Monitoring index performance

Author: The command pages have more information, so I linked to them.

   - Query profiling: use the `FT.PROFILE` command to analyze query performance and identify bottlenecks.
   - Memory usage: regularly monitor memory usage with the `INFO memory` and `FT.INFO` commands to detect growth patterns and optimize resource allocation.
   - Search query logs: enable query logging for better insights into how indexes are utilized.

#### 6. Index maintenance

>    - Reindexing: if schema changes are required, create a new index with the updated schema and reassign the alias once the index is ready.

Reviewer 2: When is it appropriate to use FT.Alter (if at all) and when is it best to use aliasing? This comes up with customers. fyi @joeywhelan

Author: I could use some example wording here.

Author: Fixed in next push.
